dispatcher.go

```diff
@@ -413,7 +555,37 @@ func (md *multicastDispatcher[T]) executeSync(streams []*asyncEventStream[T], ev
 				return
 			}
 
-			stream.stream <- evt.event
+			// Ensure that the event stream has been accessed before sending an event. This
+			// is to ensure that synchronous events do not wait for a receiver that doesn't
+			// exist, and prevents go routines from leaking.
+			if stream.accessed.Get() {
+				// When an event stream has been accessed, we also know whether or not
+				// it is synchronous. This changes the behavior of the dispatch-sync
+				// to now wait for the receiver to notify us that it is done before
+				// continuing.
+				if stream.sync.Get() {
+					syncEvent := newSyncEvent(evt.event)
+					stream.syncStream <- syncEvent
+
+					// if we included a non-zero timeout, set it here
+					if evt.timeout != 0 {
+						t := time.NewTimer(evt.timeout)
+						select {
+						// synchronous event was effectively handled and signaled
+						case <-syncEvent.done:
+							t.Stop()
+							return
+						// timeout occurred before the event could be signaled
+						case <-t.C:
+						}
```

Comment:
Why do we make a Timer both here and in the `DispathWithTimeout` method?

Reply:
Also, should the `<-t.C` case also have a return? Or should the `.done` case not have a return?

Reply:
> Why do we make a Timer both here and in the DispathWithTimeout method?

This is the reason why it's important to add the `DispatchWithTimeout` to the API itself (given that passing contexts to the event API seemed dumb to me). 

The way the plumbing for the events work to ensure that we dispatch asynchronously without blocking future event dispatching is:

1.  Field all new event dispatches on an "incoming" channel (which are received on a dedicated (per Dispatcher) go routine. 
2. Dispatch each event to the receiver (EventStream) on a separate go routine. 

Adding the timeout to DispatchSync ensures the most accurate timeout, and unblocks  DispatchSync immediately.

However, If we only timed out the DispatchSync channel, the actual dispatch go routine would leak (sit in a waiting state indefinitely) even though the `DispatchSync` call would no longer be waiting for the "sent" signal. 

Timing out both ensures that we stop blocking as soon as the timeout occurs _and_ timeout any dispatching go routines such that they don't leak.

Reply:
> Also, should the <-t.C case also have a return? Or should the .done case not have a return?

I don't think the `case <-syncEvent.done:`  requires a `return` -- we could either add one specifically to `<-t.C` or remove the return from `<-syncEvent.done`.

Reply:
Also worth noting here that I explicitly create a `Timer` here instead of using the super handy `<-time.After(duration)` because this will create a go routine that lives the length of the timeout every call. Using a timer ensures that you can call `Stop()` on a successful case, which cleans up the "timer" go routine.

Reply:
Thanks for the super detailed responses, this is a great rationale for the two timers. Would you include it in a comment for future-me?

I don't have a strong opinion in either direction about the accuracy of the timeout -- I seriously trust your gut instinct here.

Also trust your judgement on the `<-t.C` case, just wanted to make sure whatever we do is intentional.

Reply:
Removed the return for consistency!